### ISSUE TITLE
Implement a method to access the buffer rendered inside the current view for fuzzy finding

### ISSUE THREAD

**Issue by whisperity** [christianparpart/contour#168]
## Abstract

Have a way (similar to `tmux capture-pane`) to fetch the currently visible text of the current (or any other?) Contour "viewport" (not sure if that is the right word) for text processing.

## Motivation

One of my colleagues created a [Zsh script for FZF](http://github.com/martong/dotfiles/blob/46734e2a3f31b130163a9090a91a6bde1005396e/home/.fzf_tmux_hist.zsh) that through TMux (using the `tmux capture-pane` feature) allows him to do a fuzzy find on either the lines or the words rendered in the current viewport.

I've personally never used this feature as I rarely use TMux locally, but it is *somewhat* useful e.g. when you do something in the terminal and want to reuse a word (e.g. a filename?) or a line from it in a subsequent command without having to reach for the mouse.

Now, that script I linked, and the general idea, requires TMux which he uses extensively, but I don't.
However, we have an amazing terminal emulator at hand which could give us a similar "captured buffer view" for scripts to wrangle. This might depend on #149.

## Specification

No idea. Something like `contour --eval <running_session_id> --get-buffer` (where the session ID is exposed in an env var #167)  could throw back a huge textual buffer (either with all escape sequences or optionall some sort of a view of the current render only, with the escape sequences omitted) on the standard output, which then could be processed by the client. Check my colleague's script, he fetches textual data from `tmux` and then does Bash/Zsh magic on it.

## Additional context?

FWIW, I've *tried* to do something similar with my day-to-day terminal emulator of choice, Konsole. In Konsole, you can set the scrollback buffer between "None", "Fixed size n lines" and "Unlimited". From these options, "None" is obviously useless for the purposes of this issue. "Fixed size" is also useless, because then the backing data is only in RAM, and Konsole exposes no way of access to it. "Unlimited" *might* be useful as it creates the scrollback buffer at `/tmp`... which is a binary file. That was the point I gave up.

~~~~
╰─ fd 'konsole' /tmp | head -n 1 | xargs hexyl | head -n 25
┌────────┬─────────────────────────┬─────────────────────────┬────────┬────────┐
│00000000│ 63 00 00 00 01 00 00 00 ┊ 01 01 00 00 01 55 64 00 │c000•000┊••00•Ud0│
│00000010│ 00 00 01 00 00 00 01 01 ┊ 00 00 01 9c 20 00 00 00 │00•000••┊00•× 000│
│00000020│ 01 00 00 00 01 01 00 00 ┊ 01 70 45 00 00 00 01 00 │•000••00┊•pE000•0│
│00000030│ 00 00 01 01 00 00 01 40 ┊ 6e 00 00 00 01 00 00 00 │00••00•@┊n000•000│
│00000040│ 01 01 00 00 01 00 76 00 ┊ 00 00 01 00 00 00 01 01 │••00•0v0┊00•000••│
│00000050│ 00 00 01 00 70 00 00 00 ┊ 01 00 00 00 01 01 00 00 │00•0p000┊•000••00│
│00000060│ 01 00 72 00 00 00 01 00 ┊ 00 00 01 01 00 00 01 00 │•0r000•0┊00••00•0│
│00000070│ 6f 00 00 00 01 00 00 00 ┊ 01 01 00 00 01 55 62 00 │o000•000┊••00•Ub0│
│00000080│ 00 00 01 00 00 00 01 01 ┊ 00 00 01 0b 65 00 00 00 │00•000••┊00••e000│
│00000090│ 01 00 00 00 01 01 00 00 ┊ 01 a7 20 00 00 00 01 00 │•000••00┊•× 000•0│
│000000a0│ 00 00 01 01 00 00 01 00 ┊ 6c 00 00 00 01 00 00 00 │00••00•0┊l000•000│
│000000b0│ 01 01 00 00 01 55 6f 00 ┊ 00 00 01 00 00 00 01 01 │••00•Uo0┊00•000••│
│000000c0│ 00 00 01 0c 61 00 00 00 ┊ 01 00 00 00 01 01 00 00 │00•_a000┊•000••00│
│000000d0│ 01 04 64 00 00 00 01 00 ┊ 00 00 01 01 00 00 01 00 │••d000•0┊00••00•0│
│000000e0│ 65 00 00 00 01 00 00 00 ┊ 01 01 00 00 01 55 64 00 │e000•000┊••00•Ud0│
│000000f0│ 00 00 01 00 00 00 01 01 ┊ 00 00 01 0c 20 00 00 00 │00•000••┊00•_ 000│
│00000100│ 01 00 00 00 01 01 00 00 ┊ 01 01 73 00 00 00 01 00 │•000••00┊••s000•0│
│00000110│ 00 00 01 01 00 00 01 00 ┊ 75 00 00 00 01 00 00 00 │00••00•0┊u000•000│
│00000120│ 01 01 00 00 01 55 63 00 ┊ 00 00 01 00 00 00 01 01 │••00•Uc0┊00•000••│
│00000130│ 00 00 01 0c 63 00 00 00 ┊ 01 00 00 00 01 01 00 00 │00•_c000┊•000••00│
│00000140│ 01 4d 65 00 00 00 01 00 ┊ 00 00 01 01 00 00 01 00 │•Me000•0┊00••00•0│
│00000150│ 73 00 00 00 01 00 00 00 ┊ 01 01 00 00 01 55 73 00 │s000•000┊••00•Us0│
│00000160│ 00 00 01 00 00 00 01 01 ┊ 00 00 01 0d 66 00 00 00 │00•000••┊00•_f000│
│00000170│ 01 00 00 00 01 01 00 00 ┊ 01 35 75 00 00 00 01 00 │•000••00┊•5u000•0│
~~~~

Konsole exposes itself through DBus (which is beyond my expertise sadly... I'm working on vastly single-thread backend stuff :joy: #compilers), and there's a bunch of *stuff* exposed here, but none which would allow getting the history. Talking to the thing via "easy-to-use" Shell commands I believe allows much more customisation and user interfacing than doing layers upon layers of (admittedly nice, object-oriented, type-safe, etc...) bus calls.

~~~~
╰─ qdbus $KONSOLE_DBUS_SERVICE $KONSOLE_DBUS_SESSION
method QByteArray org.kde.konsole.Session.codec()
method QStringList org.kde.konsole.Session.environment()
method bool org.kde.konsole.Session.flowControlEnabled()
method int org.kde.konsole.Session.foregroundProcessId()
method int org.kde.konsole.Session.historySize()
method bool org.kde.konsole.Session.isMonitorActivity()
method bool org.kde.konsole.Session.isMonitorSilence()
method int org.kde.konsole.Session.processId()
method void org.kde.konsole.Session.runCommand(QString command)
method void org.kde.konsole.Session.sendMouseEvent(int buttons, int column, int line, int eventType)
method void org.kde.konsole.Session.sendText(QString text)
method bool org.kde.konsole.Session.setCodec(QByteArray codec)
method void org.kde.konsole.Session.setEnvironment(QStringList environment)
method void org.kde.konsole.Session.setFlowControlEnabled(bool enabled)
method void org.kde.konsole.Session.setHistorySize(int lines)
method void org.kde.konsole.Session.setMonitorActivity(bool)
method void org.kde.konsole.Session.setMonitorSilence(bool)
method void org.kde.konsole.Session.setMonitorSilenceSeconds(int seconds)
method void org.kde.konsole.Session.setTabTitleFormat(int context, QString format)
method void org.kde.konsole.Session.setTitle(int role, QString title)
method QString org.kde.konsole.Session.shellSessionId()
method QString org.kde.konsole.Session.tabTitleFormat(int context)
method QString org.kde.konsole.Session.title(int role)
method QDBusVariant org.freedesktop.DBus.Properties.Get(QString interface_name, QString property_name)
method QVariantMap org.freedesktop.DBus.Properties.GetAll(QString interface_name)
signal void org.freedesktop.DBus.Properties.PropertiesChanged(QString interface_name, QVariantMap changed_properties, QStringList invalidated_properties)
method void org.freedesktop.DBus.Properties.Set(QString interface_name, QString property_name, QDBusVariant value)
method QString org.freedesktop.DBus.Introspectable.Introspect()
method QString org.freedesktop.DBus.Peer.GetMachineId()
method void org.freedesktop.DBus.Peer.Ping()

╰─ qdbus $KONSOLE_DBUS_SERVICE $KONSOLE_DBUS_WINDOW
method int org.kde.konsole.Window.currentSession()
method QString org.kde.konsole.Window.defaultProfile()
method void org.kde.konsole.Window.moveSessionLeft()
method void org.kde.konsole.Window.moveSessionRight()
method int org.kde.konsole.Window.newSession()
method int org.kde.konsole.Window.newSession(QString profile)
method int org.kde.konsole.Window.newSession(QString profile, QString directory)
method void org.kde.konsole.Window.nextSession()
method void org.kde.konsole.Window.prevSession()
method QStringList org.kde.konsole.Window.profileList()
method int org.kde.konsole.Window.sessionCount()
method QStringList org.kde.konsole.Window.sessionList()
method void org.kde.konsole.Window.setCurrentSession(int sessionId)
method void org.kde.konsole.Window.setTabWidthToText(bool)
method QDBusVariant org.freedesktop.DBus.Properties.Get(QString interface_name, QString property_name)
method QVariantMap org.freedesktop.DBus.Properties.GetAll(QString interface_name)
signal void org.freedesktop.DBus.Properties.PropertiesChanged(QString interface_name, QVariantMap changed_properties, QStringList invalidated_properties)
method void org.freedesktop.DBus.Properties.Set(QString interface_name, QString property_name, QDBusVariant value)
method QString org.freedesktop.DBus.Introspectable.Introspect()
method QString org.freedesktop.DBus.Peer.GetMachineId()
method void org.freedesktop.DBus.Peer.Ping()
~~~

**Comment by uspasojevic96** [comment]
There is a screenshot function if you are looking for that, it saves current buffer to file

**Comment by whisperity** [comment]
@uspasojevic96 Do you mean in Konsole, or in Contour? And even for both, how do I access it from a script? What I'd like to have is similar to `:BLines` in `fzf.vim`, i.e. I press a keybind in, let's say, Zsh, and it calls back the terminal emulator (my colleague's code calls back tmux) and asks for the buffer and does some magic on it and presents me with, let's say, the lines inside it.

**Comment by christianparpart** [comment]
I think you mean you want to access the textual contents of the current viewport (the text area you currently see) or ofthe whole screen plus history lines. That is absolutely possible, but I am not sure there is a performant way to do that via the VT protocol (as an extension). There is, but will that be performant? I will have a look how tmux's capture-pane output looks like, and how it can be adapted to contour.

**Comment by christianparpart** [comment]
So here is my little drafty spec proposal:

### Syntax

It will be added as OSC VT sequence. To request the screen buffer the syntax is as follows:
```
OSC <FunctionNumber> ; Mode ; StartLine ; [Count] ST
```
With:
- `<FunctionNumber>` is yet to be determined (some free ID, such as 42, but maybe a more sensible one).
- `Mode` is either `0` or `1`. A `0` indicates that the line numbers in `StartLine` start counting from the history start with `0` value in `StartLine` being the oldest history line (and so on), whereas a value of `1` means that screen grid coordinates are used, and therefore only the main page area can be accessed between `1` and number of screen lines. (This does not necessarily correlate to what the user sees if the user scrolled back into the history).
- `StartLine` is the line number of the first line of the requested buffer (with the number being interpreted as `Mode` mandates).
- `Count` is the number of lines to be requested. A value of `0` means everything, leaving this parameter out will default to `0`.

The resulting screen buffer can be received by reading from `stdin`. The screen buffer will contain text as well as formatting VT sequences for colorization and text decoration (no cursor movements or anything else).

The buffer is complete once the following special VT sequence has been read from `stdin`.

```
OSC <FunctionNumber> ; ; ; ST
```

With `<FunctionNumber>` being the same as above.

Note, `OSC` is `\033]` and `ST` is `\033\\`.

### Security

It must not be able by an application to read out the screen if the user does not allow it. So the user must be asked in any form.
In case of contour, there is a little permission system (e.g. already used for font change requests).

### Feature Detection

Usually `DA1` the most prominent feature detection VT sequence, so I'd suggest to extend that with a new parameter that uniquely identifies this feature. That parameter should ideally match the above mentioned `<FunctionNumber>`.

### Notes

The good thing about this way is, it works across SSH connections as it is actually only extending the VT sequence stream and does not rely on any temporary file that would not be accessible when working remotely.

**Comment by whisperity** [comment]
Hrm... Okay, I'm a bit baffled and there is a lot of things in the spec I don't even understand. How would I generate that lengthy sequence *from a script file*? `tmux capture-pane` is a command that one can run from any shell and specify the "target" to it. Or can a script file just start typing these escape sequences out for the terminal emulator to pick it up? Could typing these, and the "permission asking" thing break the rendering of the stuff that's already inside the window? As far as I can see, `capture-pane` does not interfere with the targeted window in a "write" way.

**Comment by whisperity** [comment]
As for the permission thing, as long as there is a way (even if by setting a config file or giving an additional command-line flag when starting the binary) to turn the asking off so it works without having to click or type anything in, it's fine.

Naturally, the request should not allow cross-user captures. When I say `tmux list-panes` on a shared server, I do not see another user's setup, even though `ps faux` shows many other tmuces running. I think their security model here is that the socket(?) used to access the session is guarded on the filesystem. Because if I `sudo su xyz` over to their user, `tmux list-panes` shows a whole different world.

Somehow this issue could be better tied up with the daemon mode (#149) thing, I believe.

**Comment by whisperity** [comment]
So the problem with the whole thing is that if the buffer contains all the escape sequences and whatnot, there is a *lot* to be done parsing-wise on the user's side. Or.. in some sense "the user's side". It is okay if the "low-level" terminal emulator dumps "everything at you", but such an output will be tough if people have to start manually parsing them in their shell scripts and whatnot. If there would be a middle layer, like a separate command, or script, that's shipped with contour, that dumbs down the output to something usable for bare text processing, then I can imagine people starting to use this feature. Otherwise, it'll be just wasted time to implement for the average person, because if I want to "get the lines that I see in front of me", I want to get those lines, and not the low-level stuff embedded within. When you highlight with the mouse and copy the text visible, it also doesn't copy the escape sequences with it. I think `gnome-terminal` offers a *"Copy as HTML..."* feature, but that's an additional two clicks.

**Comment by christianparpart** [comment]
Dear not, i will provide an example Shell-Script that shows how to use this
VT sequence to save the current screen buffer to a local file. There will
be no magic.

Security: you can only and really only access the screen buffer of the
terminal your script is running in, by design. Everything else would be a a
security nightmare.

Permissions are configured inn the config file in contour and currently
support: permanently deny, permanently allow, as well as interactively ask
the user via popup dialog. This will not block the rendering at all but is
asynchronous in the response. The script must b aware of that (easy).

**Comment by whisperity** [comment]
Great, thanks. So in your PoV, tmux *is* a security nightmare? Because I can sure as hell fetch the rendered buffer with `capture-term` and type into the running shell with `tmux send-keys`. :joy: (The latter I had to actually look up, but I did remember `screen` having an `--exec` flag or somesuch.) Granted, I do need access to the daemon socket for this, so unless I have the ability to privilege escalate across the filesystem... at which points it's a moot argument because I can start being you in every sense of the word...

But yeah, I do not see -- for now -- the use case for accessing one terminal's buffer from another script. At least not with regards to the use cases I can imagine this "fuzzy find in the printed lines" kind of stuff.

**Comment by christianparpart** [comment]
> So in your PoV, tmux _is_ a security nightmare?

I do not tare to talk about other people's software as critically as I do with my own. ;-)

Tonic depends on proper file permissions and it is the users fault if the file permissions are too wide open. 

In case of contour i do not even want to provide that possibility. I also think there is no practical benefit of a process to push into other terminals contents. 

The whole capability should be kept simple, too.

**Comment by christianparpart** [comment]
I think we can close this now. Any further improvements can maybe go into a new ticket if necessary. :)
